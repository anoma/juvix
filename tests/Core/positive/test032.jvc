-- Ackermann function (higher-order definition)

def compose := \f \g \x g (f x);

def iterate := \f \n
  if n = 0 then
    \x x
  else
    compose f (iterate f (n - 1));

def plus := iterate (+ 1);

def mult := \m \n iterate (plus n) m 0;

def exp := \m \n iterate (mult m) n 1;

def ackermann := \m
  iterate (\f \n iterate f (n + 1) 1) m (+ 1);

def writeLn := \x write x >> write "\n";

writeLn (plus 3 7) >>
writeLn (mult 3 7) >>
writeLn (exp 3 7) >>
writeLn (ackermann 3 4)
