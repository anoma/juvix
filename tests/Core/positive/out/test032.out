10
21
2187
125
